java/memory/src/main/java/org/apache/arrow/memory/util/hash/MurmurHasher.java

```diff
@@ -0,0 +1,178 @@
+ *   integer multiplications. However, the produced hash codes have
+ *   good quality in the sense that they are uniformly distributed in the universe.
+ * </p>
+ * <p>
+ *   Therefore, this algorithm is suitable for scenarios where uniform hashing
+ *   is desired (e.g. in an open addressing hash table/hash set).
+ * </p>
+ */
+public class MurmurHasher implements ArrowBufHasher {
+
+  private final int seed;
+
+  /**
+   * Creates a default Murmur hasher, with seed 0.
+   */
+  public MurmurHasher() {
+    this(0);
+  }
+
+  /**
+   * Creates a Murmur hasher.
+   * @param seed the seed for the hasher.
+   */
+  public MurmurHasher(int seed) {
+    this.seed = seed;
+  }
+
+  @Override
+  public int hashCode(long address, int length) {
+    return hashCode(address, length, seed);
+  }
+
+  @Override
+  public int hashCode(ArrowBuf buf, int offset, int length) {
+    buf.checkBytes(offset, offset + length);
+    return hashCode(buf.memoryAddress() + offset, length);
+  }
+
+  /**
+   * Calculates the hash code for a memory region.
+   * @param buf the buffer for the memory region.
+   * @param offset offset within the buffer for the memory region.
+   * @param length length of the memory region.
+   * @param seed the seed.
+   * @return the hash code.
+   */
+  public static int hashCode(ArrowBuf buf, int offset, int length, int seed) {
+    buf.checkBytes(offset, offset + length);
+    return hashCode(buf.memoryAddress() + offset, length, seed);
+  }
+
+  /**
+   * Calculates the hash code for a memory region.
+   * @param address start address of the memory region.
+   * @param length length of the memory region.
+   * @param seed the seed.
+   * @return the hash code.
+   */
+  public static int hashCode(long address, int length, int seed) {
+    int index = 0;
+    int hash = seed;
+    while (index + 4 <= length) {
+      int intValue = getInt(address + index);
+      if (!ArrowBufPointer.LITTLE_ENDIAN) {
+        intValue = Integer.reverseBytes(intValue);
+      }
+      hash = combineHashCode(hash, intValue);
+      index += 4;
+    }
+
+    if (index < length) {
+      // process remaining data as a integer in little endian
+      int intValue = 0;
+      for (int i = index - 1; i >= index; i--) {
+        intValue <<= 8;
+        intValue |= (getByte(address + i) & 0x000000ff);
+        index += 1;
+      }
+      hash = combineHashCode(hash, intValue);
+    }
+    return finalizeHashCode(hash, length);
+  }
+
+  /**
+   * Combine the current hash code and a new int value to calculate
+   * a new hash code.
+   * @param currentHashCode the current hash code.
+   * @param intValue the new int value.
+   * @return the new hah code.
+   */
+  public static int combineHashCode(int currentHashCode, int intValue) {
+    int c1 = 0xcc9e2d51;
+    int c2 = 0x1b873593;
+    int r1 = 15;
+    int r2 = 13;
+    int m = 5;
+    int n = 0xe6546b64;
+
+    int k = intValue;
+    k = k * c1;
+    k = (k << r1) | (k >> (32 - r1));
```

Comment:
the right-shift in java is signed shift by default. the algo expects an unsigned shift.

maybe, rotateLeft(k, r1) would make this easier to read ?

Reply:
Sounds good. I have added rotateLeft and rotateRight methods.